src/rate_limiter/src/lib.rs

```diff
@@ -224,6 +231,60 @@ impl TokenBucket {
     }
 }
 
+/// State for saving a TokenBucket.
+#[derive(Versionize)]
+pub struct TokenBucketState {
+    size: u64,
+    one_time_burst: Option<u64>,
+    refill_time: u64,
+    budget: u64,
+    last_update: u64,
+    processed_capacity: u64,
+    processed_refill_time: u64,
+}
+
+impl Persist for TokenBucket {
+    type State = TokenBucketState;
+    type ConstructorArgs = ();
+    type Error = ();
+
+    fn save(&self) -> Self::State {
+        TokenBucketState {
+            size: self.size,
+            one_time_burst: self.one_time_burst,
+            refill_time: self.refill_time,
+            budget: self.budget,
+            // When saving last_update, we record the number of elapsed nanoseconds
+            // since the last update took place.
+            last_update: self.last_update.elapsed().as_nanos() as u64,
+            processed_capacity: self.processed_capacity,
+            processed_refill_time: self.processed_refill_time,
+        }
+    }
+
+    fn restore(_: Self::ConstructorArgs, state: &Self::State) -> Result<Self, Self::Error> {
+        // When restoring last_update, we create a new Instant::now(), and then subtract the
+        // nanos elapsed since the last update took place, which were captured during save().
+        let elapsed = state.last_update;
+        let now = Instant::now();
+        let last_update = if let Some(instant) = now.checked_sub(Duration::from_nanos(elapsed)) {
+            instant
+        } else {
+            now
+        };
+
+        Ok(TokenBucket {
+            size: state.size,
+            one_time_burst: state.one_time_burst,
+            refill_time: state.refill_time,
+            budget: state.budget,
+            last_update,
+            processed_capacity: state.processed_capacity,
+            processed_refill_time: state.processed_refill_time,
+        })
+    }
+}
+
```

Comment:
I believe we should avoid the `TokenBucketState` manual code and derive `Versionize` directly on `TokenBucket`. It's composed of primitives and it's very well contained, it does not rely on any external elements.

Reply:
Do we consider `Instant` to be a primitive? there is no straight forward way to serialize it, we have to provide some methods that depend on the meaning of the `Instant`, in this case as a last_update field.

Reply:
How about we first decide on the guidelines for ser/de so that we can have a uniform implementation for all devices/other things that have state? We can also talk about exceptions to the rule, which I believe is something you are talking about here. We have a sync in a few hours.

Reply:
Indeed. Manual translation is required here with or without using an `Instant` object.

Whichever way we try to represent that `Instant`, the actual value still **has to be post-processed** to take into consideration the time elapsed between serialization and deserialization.

Disregard my original comment.